spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConvertersConfigurer.java

```diff
@@ -0,0 +1,121 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.cloud.stream.binding;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+import org.springframework.aop.framework.Advised;
+import org.springframework.aop.support.AopUtils;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cloud.stream.config.BindingProperties;
+import org.springframework.cloud.stream.config.ChannelBindingServiceProperties;
+import org.springframework.cloud.stream.converter.AbstractFromMessageConverter;
+import org.springframework.cloud.stream.converter.ByteArrayToStringMessageConverter;
+import org.springframework.cloud.stream.converter.CompositeMessageConverterFactory;
+import org.springframework.cloud.stream.converter.JavaToSerializedMessageConverter;
+import org.springframework.cloud.stream.converter.JsonToPojoMessageConverter;
+import org.springframework.cloud.stream.converter.JsonToTupleMessageConverter;
+import org.springframework.cloud.stream.converter.MessageConverterUtils;
+import org.springframework.cloud.stream.converter.PojoToJsonMessageConverter;
+import org.springframework.cloud.stream.converter.PojoToStringMessageConverter;
+import org.springframework.cloud.stream.converter.SerializedToJavaMessageConverter;
+import org.springframework.cloud.stream.converter.StringToByteArrayMessageConverter;
+import org.springframework.cloud.stream.converter.TupleToJsonMessageConverter;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.integration.channel.AbstractMessageChannel;
+import org.springframework.messaging.converter.MessageConverter;
+import org.springframework.util.Assert;
+import org.springframework.util.MimeType;
+import org.springframework.util.StringUtils;
+
+/**
+ * Class that configures message converters per channel binding.
+ *
+ * @author Ilayaperumal Gopinathan
+ */
+public class MessageConvertersConfigurer implements ApplicationContextAware, InitializingBean {
+
+	private static final String CUSTOM_MSG_CONVERTERS_BEAN_NAME = "customMessageConverters";
+
+	@Autowired
+	private ChannelBindingServiceProperties channelBindingServiceProperties;
+
+	private ApplicationContext applicationContext;
+
+	private CompositeMessageConverterFactory messageConverterFactory;
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		Collection<AbstractFromMessageConverter> messageConverters = new ArrayList<AbstractFromMessageConverter>();
+		if (applicationContext.containsBean(CUSTOM_MSG_CONVERTERS_BEAN_NAME)) {
+			messageConverters.addAll(
+					(Collection<AbstractFromMessageConverter>) applicationContext.getBean(CUSTOM_MSG_CONVERTERS_BEAN_NAME));
```

Comment:
@artembilan my understand was that the `beanFactory` lookup should end up creating the `singleton` as long as the `BeanDefinition` for that name exists. Though I see your point, I would like to take this opportunity to understand the side effects of early instantiation of the `singleton` this way. If you have any pointers please share. Meanwhile, I will switch to use `SmartInitializingSingleton`.